substratevm/src/com.oracle.svm.core.graal.llvm/src/com/oracle/svm/core/graal/llvm/util/LLVMIRBuilder.java

```diff
@@ -62,14 +68,28 @@ public class LLVMIRBuilder implements AutoCloseable {
     private LLVMBuilderRef builder;
     private LLVMModuleRef module;
     private LLVMValueRef function;
+    public LLVMDIBuilderRef diBuilder;
 
+    // Maps filenames to a compile unit
+    public HashMap<String, LLVMMetadataRef> diFilenameToCU;
```

Comment:
I think LLVM automatically merges identical metadata nodes when it encounters them, if that's the case we can probably get rid of those caches. Also, shouldn't all the debug information from a function have the same file name?

Reply:
I'm not sure what the answer is to that last question. However, I'd just note that a compiled method can include code from inlined methods which belong to a different source file.

Reply:
From my understanding, LLVM automatically merges identical nodes during linking. ("...duplicate information is automatically merged by the linker" https://llvm.org/docs/SourceLevelDebugging.html)  But that means the individual bytecode files corresponding to the functions(e.g. `fXX.bc`) will have duplicate compile units and subprograms without the caches based on how the code is written. So the user will have to deal with duplicate debug info if they wish to perform static analysis on the function level bytecode files. Atleast, for my use case, I would like to perform static analysis on individual bytecode files which is why I added these caches, when I saw that there were multiple compile units present in the `fXX.bc` files. 

If you think that is not a general use case, then I can remove the caches.